source/MaterialXTest/ShaderGen.cpp

```diff
@@ -40,6 +41,98 @@ bool isTopologicalOrder(const std::vector<const mx::SgNode*>& nodeOrder)
     return true;
 }
 
+// Observer to add library markers to all nodes. Use the "namespace" attribute to do this
+static std::string NAMESPACE_ATTRIBUTE("namespace");
+class LibraryObserver : public MaterialX::Observer
+{
+public:
+    LibraryObserver() {}
+
+    void onAddElement(MaterialX::ElementPtr /*parent*/, MaterialX::ElementPtr element) override
+    {
+        if (_libraryRefName.length())
+        {
+            element->setAttribute(NAMESPACE_ATTRIBUTE, _libraryRefName);
+        }
+    }
+
+    void setLibraryRefName(const std::string& name)
+    {
+        _libraryRefName = name;
+    }
+
+protected:
+    // Libraryref name to add
+    std::string _libraryRefName;
+};
+
+//
+// Get source content, source path and resolved paths for
+// an implementation
+//
+bool getShaderSource(MaterialX::ShaderGeneratorPtr generator,
+                    const MaterialX::ImplementationPtr implementation,
+                    MaterialX::FilePath& sourcePath,
+                    MaterialX::FilePath& resolvedPath,
+                    std::string& sourceContents) 
+{
+    if (implementation)
+    {
+        sourcePath = implementation->getFile();
+        resolvedPath = generator->findSourceCode(sourcePath);
+        if (MaterialX::readFile(resolvedPath.asString(), sourceContents))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+//
+// Find the implementation for specific language or target for a given Element
+//
+// Should this be moved under the document class?
+MaterialX::ImplementationPtr findImplementation(mx::DocumentPtr document, const std::string& name,
+                                                const std::string& language, const std::string& target)
+{
+    std::vector<MaterialX::InterfaceElementPtr> impllist = document->getMatchingImplementations(name);
+    for (auto elem : impllist)
+    {
+        MaterialX::ImplementationPtr implementation = elem->asA<MaterialX::Implementation>();
+        if (implementation)
+        {
+            // Check for a language match
+            const std::string& lang = implementation->getLanguage();
+            if (lang.length() && lang == language)
+            {
+                // Check target. Note that it may be empty.
+                const std::string& targ = implementation->getTarget();
+                if (!targ.length() || targ == target)
+                {
+                    return implementation;
+                }
+            }
+        }
+    }
+    return nullptr;
+}
+
+// Check if a nodedef requires an implementation check
+// Untyped nodes do not.
+bool requiresImplementation(const MaterialX::NodeDefPtr nodeDef) 
+{
+    if (!nodeDef)
+        return false;
+
+    static std::string TYPE_NONE("none");
+    const std::string typeAttribute = nodeDef->getAttribute("type");
+    if (typeAttribute.length() == 0 || typeAttribute == TYPE_NONE)
+    {
+        return false;
+    }
+    return true;
```

Comment:
Could be simplified to: `return !typeAttribute.empty() && typeAttribute != TYPE_NONE;`